### PR TITLE
Add new version of rowWilcoxonTest

### PR DESCRIPTION
**Pull request by nicolas-enjalbert**
(keep last version and add V1 (soft loop of permutation) and V3 (matrix optimization)

### CODE REVIEW

**Comment by pneuvial** [comment]
Thanks, this is great. Can you also add tests to `tests/testthat/`

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:85]: Here we implicitly assume that the number of 0 and 1 items is the same for all columns of 'categ'. This is indeed the case if `categ` is generated by permutations of an initial vector. We should raise an error if it's not the case, something like:

```suggestion
    nx <- sum(categ[,1] == 1) 
    nxs <- colSums(categ)
    if (any(nxs != nx)) {
      stop("All columns of 'categ' should have the same number of 0 and 1")
    }
    n <- nrow(categ)
    ny <- n - nx
```

Then we should update the corresponding test to check that an error is indeed raised in this case.

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:138]: Same comment as for `rowWilcoxonTests` (the check may be here or in the loop on `b` below).

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:127]: Reading this version of the code I now think that we should probably drop this version entirely (ie drop`rowWilcoxonTestsV2` and `rowWilcoxonTests1V2`):  it is "au milieu du gué", with the `rowWilcoxonTests1V2` having as arguments both `categ` and `nx` and `ny`,  which is awkward and redundant.

File: R/rowWilcoxonTests.R
Comment (pneuvial) [reply R/rowWilcoxonTests.R:138]: This comment should be ignored if the entire function is dropped!

**Review by pneuvial — CHANGES_REQUESTED** [review]
Thanks! I think it's already close to be mergeable.

File: R/rowWilcoxonTests.R
Comment (pneuvial) [reply R/rowWilcoxonTests.R:85]: Elaborating on the above comment:

The original version of `rowWilcoxonTests` would actually work if the number of 1 in the `categ` matrix was not the same for each columns.  It is not the case anymore, because `nx` and `ny` are scalars and not vectors of length `B`. 

We should make sure that this case is still covered!

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:95]: This should work when `nx` and `ny` are vectors and not scalars.

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:96]: Here one should be extra careful when `stats`  is a matrix and `nx` and `ny` are vectors!
I guess something like 

```
sweep(stats, MARGIN = 2, STATS = ny * nx / 2, FUN = "-")
```

should do the trick?

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:103]: Here extra care is needed too for `/`.

**Review by pneuvial — COMMENTED** [review]
More comments in order to cover the case when the columns of `categ` do not have the same number of 0 and 1.

**Comment by nicolas-enjalbert** [comment]
This version covers the case when `categ` is not only a permutation of the vector `groups`.

File: R/rowWilcoxonTests.R
Comment (pneuvial) [inline R/rowWilcoxonTests.R:102]: This is probably not optimal: the formula is of the form t(a - M/b), where a and b are vectors and M is a matrix. So it would be safer to have the "-" and "/" replaced by 'sweep' operations. I leave it to you to decide whether you want to change this now.

**Comment by pneuvial** [comment]
Thanks! I'm OK with merging (see last comment about sigma above).

**Comment by nicolas-enjalbert** [comment]
I added this new version using sweep method instead of vectorisation method for sigma. This version is safer for the maintainer. Nevertheless, I found the vectorisation method is faster than the sweep method in average, 422 milliseconds against 1033 (for B = 1000, nb_exp = 100)

**Review by pneuvial — APPROVED** [review]

**Comment by pneuvial** [comment]
Great. Can you also add a script inst/system/speed_rowWilcoxonTests.R which reproduces some of the plots you showed me on performance improvement compared to wilcox.test ?

**Pull request merged**
